### PR TITLE
Racer do lighthouse

### PR DESCRIPTION
**Pull request by spember**
## Description
This PR adds:
* basic Lighthouse execution (note that it does _not_ yet accept config, e.g. additional chrome flags. This will be a follow up
* an in-memory repository of the lighthouse results
* an endpoint to kick off the lighthouse run
* a locking mechanism to ensure that multiple runs cannot be executed

if you want to test this out, checkout and run:
`yarn build && docker-compose up --build`

then, from another tab:

`curl "localhost:3000/race" -X POST -d '{"targetUrl":"https://www.google.com"}'` or whichever target url

## Type of Change

- [ ] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

### CODE REVIEW

File: packages/racer/src/usageLock.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Represents a Mutex or Lock used to track if Lighthouse is running or not.
```
Comment (roippi) [inline packages/racer/src/usageLock.ts:2]: Can we use something off the shelf for this?  e.g. https://www.npmjs.com/package/async-mutex

File: packages/racer/src/profiling/index.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    // screenEmulation: userConfig.screenEmulation,
+  };
+  launch(chromeOptions).then((chrome) => {
+    const lhFlags = {
+      //   chromeOptions.chromeFlags,
+      port: chrome.port,
+      // unfortunately, setting a max wait causes the lighthouse run to break. can investigate in the future
+      // maxWaitForLoad: 12500
+    };
+    return new Promise((resolve, reject) => {
```
Comment (roippi) [inline packages/racer/src/profiling/index.ts]: mixing `async/await` styles from [profile.ts](https://github.com/wayfair-incubator/racepoint/pull/5/files#diff-d0fbb35c60deeffd39623441e7fac394c1b21e357aa7df61fded31e971586544R51) and using promise.then/return new Promise here.  Can we go all-in on one style?  (I don't prefer one or the other)

**Review by roippi — COMMENTED** [review]
Couple of comments, looking good

File: packages/racer/src/profiling/index.ts (hunk 2 above)
Comment (spember) [reply packages/racer/src/profiling/index.ts]: Fair call out, I'll do a pass and clean it up. After thinking about it a bit, I'd like to try to use async/await syntax as much as possible. I prefer the functional style, yet I believe that async/await will be more approachable to others (e.g. if this thing gets big and outsiders come in).

File: packages/racer/src/usageLock.ts (hunk 1 above)
Comment (spember) [reply packages/racer/src/usageLock.ts:2]: yeah, I looked at that one; even read through some of the source. Like I mentioned, the library is quite powerful and beyond what we need at this moment. What I have here does work, and adopting a library may take additional time to ingest and integrate. I will happily switch to it if my little lock thing proves faulty or we need more power!

File: packages/racer/src/controllers/common.ts [hunk 3]
```diff
@@ -1,15 +1,43 @@
...
+  method: string;
   handler: (
     request: IncomingMessage,
     response: ServerResponse,
     url: url.UrlWithParsedQuery
-  ) => Promise<any>;
+  ) => Promise<EndpointResponse<T>>;
+}
+
+export class EndpointResponse<T> {
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/controllers/common.ts:16]: Might be clearer as:
```
export interface EndpointContext {}

export class EndpointResponse<EC extends EndpointContent> {
```
or similar

File: packages/racer/src/controllers/profile.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+const isValid = (context: RaceContext): boolean => {
+  //trivial validation for now
+  return context != undefined && context.targetUrl != undefined;
+};
+
+const maybeRunLighthouse = async (
+  context: RaceContext
+): Promise<EndpointResponse<object>> => {
+  // try to get a lock; if LH is currently in use, send back a 503!
+  if (await UsageLock.getInstance().tryAcquire()) {
```
Comment (JasonPapanicholasWF) [inline packages/racer/src/controllers/profile.ts:30]: I would think that tryAcquire would return false if it was not in use 🤔

Comment (spember) [reply packages/racer/src/controllers/profile.ts:30]: think of it as 'true -> you can go ahead'. you have successfully acquired the lock

File: packages/racer/src/controllers/common.ts (hunk 3 above)
Comment (spember) [reply packages/racer/src/controllers/common.ts:16]: so put the headers and status code in a Context? would the body go in the context as well or be part of the Response?
In a way, the 'response' is itself a 'context': basically telling the server "here's bunch of info encapsulating the result of what this endpoint was supposed to do".

**Pull request merged**